### PR TITLE
Fix hooks

### PR DESCRIPTION
**Pull request by iilei**
* Add tests for Hooks execution order and amount
* follow-up on #140

### CODE REVIEW

**Comment by iilei** [comment]
I altered the integration test so it checks fragments serving fewer javascript files than permitted by `maxAssetLink` option and also added a test that checks execution of hooks.

File: lib/fragment.js [hunk 1]
```diff
@@ -183,9 +184,12 @@ module.exports = class Fragment extends EventEmitter {
             return;
         }
 
+        const range = [this.index, this.index + this.scriptRefs.length - 1];
+        const fragmentId = this.fragmentId;
+
         this.scriptRefs.forEach((uri)=> {
             const id = this.index;
-            const attributes = Object.assign({}, this.getPipeAttributes(), { id: this.attributes.id || id });
+            const attributes = Object.assign({}, this.getPipeAttributes(), { id: this.attributes.id || id, range, fragmentId });
```
Comment (vigneshshanmugam) [inline lib/fragment.js]: + We should combine both id and fragmentId 
+ Use index only for script ordering
+ Id/fragmentId for all the other cases - context, perf hooks

Comment (vigneshshanmugam) [inline lib/fragment.js]: Check here https://github.com/zalando/tailor#fragment-attributes

**Comment by vigneshshanmugam** [comment]
There is an interesting bug here, If a fragment assume `fragment-1` sends three scripts in Link header
```html
Link: <module-1.js>; rel="fragment-script", <module-2.js>; rel="fragment-script", <module-3.js>; rel="fragment-script";
```
1. Module-1 exposes a default export function 
2. Module-2 does not exposes any export function, Instead just executes some arbitrary stuff. 
3. Same as module-1

In this scenario, The hooks except `start` for the fragment-1 will never run and `all-done` will not run as well. 

This was not a problem before because we ignore if the fragment does not expose any default exports and fire all-done at last.

**Comment by iilei** [comment]
Good catch ... I haven't used require.js before and I wonder if there is some callback that would be able to handle the load event of arbitrary stuff.

**Comment by vigneshshanmugam** [comment]
Oops Sadly, this was broken before as well.. 'all-done' was never called if there was a script without a default export. I will try to find some hooks for require.js in the meanwhile.

**Comment by iilei** [comment]
I am also trying to find a way on another branch 'fix_hooks_wip'.
that approach would be less staight forward compared to the theoretical solution incorporating some kind of require callback, though.
let's see what works best in the end.

**Comment by vigneshshanmugam** [comment]
We need to handle three cases

1. if the script is not a AMD module, callback wont even get called
2. if script is AMD and does not have a init function(standalone)
3. if script is AMD and exposes a init fn (handled already - default)

**Comment by vigneshshanmugam** [comment]
I am starting to think, this logic should not be in tailor

1. Tailor should just fire onStart, onBefore, onAfter for all fragment scripts.
2. onDone should be called when all fragment's scripts are done initialising 
3. Whoever wants to hook in to the logic should group them based on fragment Ids or whatever logic. We already handle the timingGroup feature in similar fashion https://github.com/zalando/tailor/blob/master/examples/fragment-performance/templates/index.html#L7

Thoughts? @iilei

**Comment by vigneshshanmugam** [comment]
I will create a new PR with your commits and some of my own.

**Comment by iilei** [comment]
I agree. That would be a good way of separating concerns and reducing the footprint of tailor. therefore :thumbsup:

File: src/pipe.js [hunk 2]
```diff
@@ -63,10 +73,12 @@
         start.parentNode.removeChild(start);
         end.parentNode.removeChild(end);
         script && require([script], function (i) {
-            // Exposed fragment initialization Function/Promise
+            // Exported AMD fragment initialization Function/Promise
             var init = i && i.__esModule ? i.default : i;
             // early return
             if (typeof init !== 'function') {
+                initState.pop();
```
Comment (vigneshshanmugam) [inline src/pipe.js]: This will fix the issue when onDone was not called for script without dependencies.

File: lib/fragment.js [hunk 3]
```diff
@@ -197,17 +198,18 @@ module.exports = class Fragment extends EventEmitter {
      * Insert the placeholder for pipe assets at the end of fragment stream
      */
     insertEnd () {
-        this.index--;
         if (this.scriptRefs.length > 0) {
+            const range = [this.index - this.scriptRefs.length, this.index - 1];
+            this.index--;
+            const fragmentId = this.attributes.id || range[0];
```
Comment (vigneshshanmugam) [inline lib/fragment.js]: By keeping the id consistent, It ll become easy to measure the fragment perf. I will update the fragment-performance script to handle this case.

**Comment by vigneshshanmugam** [comment]
@iilei Its ready, have added some changes to your PR. Let me know your thoughts, We can merge once you review it.

**Comment by iilei** [comment]
@vigneshshanmugam great stuff :+1: 

I just added implicit testing the amount of hook execution by appending semicolons to the log lines.

**Comment by vigneshshanmugam** [comment]
@iilei I did manually remove it :) All good

**Comment by vigneshshanmugam** [comment]
Final rebase done. Merging it

**Pull request merged**